数据清洗

In [2]:
import pandas as pd
import numpy as np
import pyecharts.options as opts
from pyecharts.charts import Line,Grid,Bar,Pie,Funnel,Page,Tab

In [3]:
#数据加载
data_df = pd.read_csv( './projct_dataset.csv ')

In [4]:
#数据预览
data_df.head()

user_id    item_id  behavior_type user_geohash  item_category  \
0  98047837  232431562              1          NaN           4245   
1  97726136  383583590              1          NaN           5894   
2  98607707   64749712              1          NaN           2883   
3  98662432  320593836              1      96nn52n           6562   
4  98145908  290208520              1          NaN          13926   

            time  
0  2018-12-06 02  
1  2018-12-09 20  
2  2018-12-18 11  
3  2018-12-06 10  
4  2018-12-16 21

In [5]:
data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12256906 entries, 0 to 12256905
Data columns (total 6 columns):
 #   Column         Dtype 
---  ------         ----- 
 0   user_id        int64 
 1   item_id        int64 
 2   behavior_type  int64 
 3   user_geohash   object
 4   item_category  int64 
 5   time           object
dtypes: int64(4), object(2)
memory usage: 561.1+ MB


In [6]:

#检查重复值
data_df.duplicated().sum()

4092866

In [7]:
#删除重复值
data_df.drop_duplicates(inplace=True)

In [8]:
#删除无用列
data_df.drop(columns=[ 'user_geohash'], inplace=True)
data_df.shape

(8164040, 5)

In [9]:
#检查缺失值
data_df.isnull().sum()

user_id          0
item_id          0
behavior_type    0
item_category    0
time             0
dtype: int64

In [10]:
#预览潸洗后的数据集
data_df.head()

user_id    item_id  behavior_type  item_category           time
0  98047837  232431562              1           4245  2018-12-06 02
1  97726136  383583590              1           5894  2018-12-09 20
2  98607707   64749712              1           2883  2018-12-18 11
3  98662432  320593836              1           6562  2018-12-06 10
4  98145908  290208520              1          13926  2018-12-16 21

数据处理

In [11]:
#添加行为类型描述列
behavior_type_dict = {1:'点击',2:'收藏',3:'加入购物车',4:'支付'}
data_df['behavior_desc'] = data_df['behavior_type'].map(behavior_type_dict)
data_df.head()

user_id    item_id  behavior_type  item_category           time  \
0  98047837  232431562              1           4245  2018-12-06 02   
1  97726136  383583590              1           5894  2018-12-09 20   
2  98607707   64749712              1           2883  2018-12-18 11   
3  98662432  320593836              1           6562  2018-12-06 10   
4  98145908  290208520              1          13926  2018-12-16 21   

  behavior_desc  
0            点击  
1            点击  
2            点击  
3            点击  
4            点击

In [12]:
#处理时间列
data_df[['date','hour']]= data_df['time'].str.split(' ',expand=True)

In [13]:
data_df.head()

user_id    item_id  behavior_type  item_category           time  \
0  98047837  232431562              1           4245  2018-12-06 02   
1  97726136  383583590              1           5894  2018-12-09 20   
2  98607707   64749712              1           2883  2018-12-18 11   
3  98662432  320593836              1           6562  2018-12-06 10   
4  98145908  290208520              1          13926  2018-12-16 21   

  behavior_desc        date hour  
0            点击  2018-12-06   02  
1            点击  2018-12-09   20  
2            点击  2018-12-18   11  
3            点击  2018-12-06   10  
4            点击  2018-12-16   21

In [14]:
#更改数据类型
data_df['date'] = pd.to_datetime(data_df["date"])
data_df['hour'] = data_df['hour'].astype('int')

In [15]:
data_df.head()

user_id    item_id  behavior_type  item_category           time  \
0  98047837  232431562              1           4245  2018-12-06 02   
1  97726136  383583590              1           5894  2018-12-09 20   
2  98607707   64749712              1           2883  2018-12-18 11   
3  98662432  320593836              1           6562  2018-12-06 10   
4  98145908  290208520              1          13926  2018-12-16 21   

  behavior_desc       date  hour  
0            点击 2018-12-06     2  
1            点击 2018-12-09    20  
2            点击 2018-12-18    11  
3            点击 2018-12-06    10  
4            点击 2018-12-16    21

In [16]:
data_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8164040 entries, 0 to 12256904
Data columns (total 8 columns):
 #   Column         Dtype         
---  ------         -----         
 0   user_id        int64         
 1   item_id        int64         
 2   behavior_type  int64         
 3   item_category  int64         
 4   time           object        
 5   behavior_desc  object        
 6   date           datetime64[ns]
 7   hour           int32         
dtypes: datetime64[ns](1), int32(1), int64(4), object(2)
memory usage: 529.4+ MB


In [17]:
data_df['date'].min()

Timestamp('2018-11-18 00:00:00')

In [18]:
data_df['date'].max()

Timestamp('2018-12-18 00:00:00')

In [19]:
#按时间排序
data_df = data_df.sort_values(by=['date','hour'])
data_df.head(20)

user_id    item_id  behavior_type  item_category           time  \
2577   112707614  343080076              1          13230  2018-11-18 00   
3000   112707614  346570272              1           5689  2018-11-18 00   
5691   116101597   70407447              1           6512  2018-11-18 00   
5727   116101597  327916552              1           6512  2018-11-18 00   
6811   117903708  133429705              1          10725  2018-11-18 00   
6892   117903708  182627801              1          10725  2018-11-18 00   
6964   117903708   10979028              1          10725  2018-11-18 00   
7099   117903708   28793293              1          10725  2018-11-18 00   
7270   117903708  225462017              1          10725  2018-11-18 00   
7351   117903708  116692968              1          10725  2018-11-18 00   
7477   117903708  305176450              1          10725  2018-11-18 00   
7504   117903708  103910785              1          10725  2018-11-18 00   
8125   117903708  124704175              1           9175  2018-11-18 00   
8674   117903708  316815833              1           9836  2018-11-18 00   
8800   117903708  395200199              1           1807  2018-11-18 00   
18759   27049458    9984637              1           1843  2018-11-18 00   
19053   28962754  342382100              1           6648  2018-11-18 00   
19098   28962754  338038712              1           6648  2018-11-18 00   
19155   27049458   16072931              1           9820  2018-11-18 00   
19173   27049458  229197097              1          11552  2018-11-18 00   

      behavior_desc       date  hour  
2577             点击 2018-11-18     0  
3000             点击 2018-11-18     0  
5691             点击 2018-11-18     0  
5727             点击 2018-11-18     0  
6811             点击 2018-11-18     0  
6892             点击 2018-11-18     0  
6964             点击 2018-11-18     0  
7099             点击 2018-11-18     0  
7270             点击 2018-11-18     0  
7351             点击 2018-11-18     0  
7477             点击 2018-11-18     0  
7504             点击 2018-11-18     0  
8125             点击 2018-11-18     0  
8674             点击 2018-11-18     0  
8800             点击 2018-11-18     0  
18759            点击 2018-11-18     0  
19053            点击 2018-11-18     0  
19098            点击 2018-11-18     0  
19155            点击 2018-11-18     0  
19173            点击 2018-11-18     0

In [20]:
#保存清洗处理的数据集
data_df.to_csv('./cln_proj_dataset.csv', index=False, encoding='utf-8')

子图制作

In [21]:
#加载处理的数据集
data_df = pd.read_csv('./cln_proj_dataset.csv')
data_df['date'] = pd.to_datetime(data_df['date'])
data_df.head()

user_id    item_id  behavior_type  item_category           time  \
0  112707614  343080076              1          13230  2018-11-18 00   
1  112707614  346570272              1           5689  2018-11-18 00   
2  116101597   70407447              1           6512  2018-11-18 00   
3  116101597  327916552              1           6512  2018-11-18 00   
4  117903708  133429705              1          10725  2018-11-18 00   

  behavior_desc       date  hour  
0            点击 2018-11-18     0  
1            点击 2018-11-18     0  
2            点击 2018-11-18     0  
3            点击 2018-11-18     0  
4            点击 2018-11-18     0

流量指标

In [22]:
#日流量指标
#日PV
daily_pv = data_df[data_df['behavior_desc']=='点击'].groupby('date').count()['user_id'].reset_index( ).rename(columns ={'user_id': 'pv'})
daily_pv.head()

date      pv
0 2018-11-18  215480
1 2018-11-19  213108
2 2018-11-20  206740
3 2018-11-21  196121
4 2018-11-22  213292

In [23]:
#日uv
daily_uv = data_df[data_df['behavior_desc']=='点击'].groupby('date').nunique()['user_id'].reset_index( ).rename(columns ={'user_id': 'uv'})
daily_uv.head()

date    uv
0 2018-11-18  6340
1 2018-11-19  6418
2 2018-11-20  6332
3 2018-11-21  6275
4 2018-11-22  6184

In [39]:
#可视化日PV
daily_pv_line = (
        Line(init_opts=opts.InitOpts(width='1000px', height='500px'))
        .add_xaxis(xaxis_data=daily_pv['date'].tolist( ))
        .add_yaxis(
            series_name="日pv",
            y_axis=daily_pv['pv'].tolist(),
            is_symbol_show=True,
            label_opts=opts.LabelOpts(is_show=False)
        )
        .extend_axis(
            yaxis=opts.AxisOpts(
                name="日uv",
                type_="value",
                min_=0,
                max_=8000
            )
        )
        .set_global_opts(
            title_opts=opts.TitleOpts(
                title='用户日流量指标'),
            tooltip_opts=opts.TooltipOpts(
                trigger='axis',
                axis_pointer_type= 'line'),
                xaxis_opts=opts.AxisOpts(
                type_='time',
                name='日期',
            ),
                yaxis_opts=opts.AxisOpts(
                type_='value',
                name='日pv')
    )
)
daily_uv_line = (
    Line()
    .add_xaxis(xaxis_data=daily_uv['date'].tolist())
    .add_yaxis(
        series_name="日uv",
        y_axis=daily_uv['uv'].tolist(),
        is_symbol_show=True,
        label_opts=opts.LabelOpts(is_show=False),
        yaxis_index=1
    )
)
daily_pv_line.overlap(daily_uv_line)
daily_pv_line.render_notebook()


观察结果:在该月内，访问量和用户量整体缓慢攀升，并在双十二期间达到峰值，可见双十二的活动效果明显。

In [25]:
#小时流量指标
#小时PV
hourly_pv = data_df[data_df['behavior_desc']=='点击'].groupby('hour').count()['user_id'].reset_index().rename(columns ={'user_id': 'pv'})
hourly_pv.head()

hour      pv
0     0  235979
1     1  123931
2     2   68462
3     3   46794
4     4   37445

In [26]:
#小时uv
hourly_uv = data_df[data_df['behavior_desc']=='点击'].groupby('hour').nunique()['user_id'].reset_index().rename(columns ={'user_id': 'uv'})
hourly_uv.head()

hour    uv
0     0  5750
1     1  3747
2     2  2513
3     3  1916
4     4  1751

In [27]:
#可视化小时PV
hourly_pv_line = (
        Line(init_opts=opts.InitOpts(width='1000px', height='500px'))
        .add_xaxis(xaxis_data=hourly_pv['hour'])
        .add_yaxis(
            series_name="小时pv",
            y_axis=hourly_pv['pv'].tolist(),
            is_symbol_show=True,
            label_opts=opts.LabelOpts(is_show=False)
        )
        .extend_axis(
            yaxis=opts.AxisOpts(
                name="小时uv",
                type_="value",
                min_=0,
                max_=10000
            )
        )
        .set_global_opts(
            title_opts=opts.TitleOpts(
                title='用户小时流量指标'),
            tooltip_opts=opts.TooltipOpts(
                trigger='axis',
                axis_pointer_type= 'line'),
            yaxis_opts=opts.AxisOpts(
                type_='value',
                name='小时pv')
    )
)
hourly_uv_line = (
    Line()
    .add_xaxis(xaxis_data=hourly_uv['hour'].tolist())
    .add_yaxis(
        series_name="小时uv",
        y_axis=hourly_uv['uv'].tolist(),
        is_symbol_show=True,
        label_opts=opts.LabelOpts(is_show=False),
        yaxis_index=1
    )
)
hourly_pv_line.overlap(hourly_uv_line)
hourly_pv_line.render_notebook()

观察结果:23点到次日5点的访问量和用户量都在下降，5点到18点回升到相对稳定的水平:18点到22点访问量有比较明显的提升，并在21点和22点达到峰值,而用户量的趋势类似，但相对不那么明显。
小结: 18-22点为用户活跃时段，5-18点为用户一般活跃时段，23-5点为用户非活跃时段。

用户购买次数分布

In [28]:
#用户购买次数分布
user_buy_time = data_df[data_df['behavior_desc'] =='支付'].groupby('user_id').size()
buy_hist, bins = np.histogram(user_buy_time.values, bins=10,range=[0, 100])
buy_hist

array([5061, 2207,  829,  358,  218,   98,   45,   15,   11,   15],
      dtype=int64)

In [29]:
buy_time_category = [str(int(a)) + '-' + str(int(b)) for a, b in zip(bins[:-1],bins[1:])]
buy_time_category

['0-10',
 '10-20',
 '20-30',
 '30-40',
 '40-50',
 '50-60',
 '60-70',
 '70-80',
 '80-90',
 '90-100']

In [32]:
#可视化直方图结果
buy_hist_bar = (
    Bar(init_opts=opts.InitOpts(width='800px', height=' 500px'))
    .add_xaxis(buy_time_category)
    .add_yaxis(
        series_name="",
        y_axis=buy_hist.tolist(),
        category_gap=0.2
    )
    .set_global_opts(
        title_opts=opts.TitleOpts(
            title='用户购买次数分布',
        ),
    )
)
buy_hist_bar.render_notebook()

观察结果:从购买次数分布图来看，购买次数20次以内的，尤其是10次以内的用户比例非常大。
小结:应重点关注购买次数在20次以内，尤其是10次以内的用户，进一步挖掘其购买力。

热力商品和类别分析

In [33]:
#点击量排名前10的商品种类
top_item_categories = data_df.groupby('item_category')['user_id'].count().sort_values(ascending=False).head(10)
top_item_categories.name = 'Count'
top_item_cat_df = top_item_categories.to_frame().reset_index()
top_item_cat_df

item_category   Count
0           1863  264030
1          13230  231533
2           5027  225397
3           5894  217677
4           6513  196461
5           5399  191046
6          11279  123670
7           2825  108429
8           5232   96138
9          10894   92577

In [34]:
top_cat_pie = (
    Pie(init_opts=opts.InitOpts(width='500px'))
    .add("", top_item_cat_df.values.tolist())
    .set_global_opts(
        title_opts=opts.TitleOpts(title="Top10商品类别"),
        legend_opts=opts.LegendOpts(is_show=False),
    )
    .set_series_opts(label_opts=opts.LabelOpts(formatter="{d}%"))
)
top_cat_pie.render_notebook()

In [35]:
#点击量排名前10的商品
top_items = data_df.groupby('item_id')['user_id'].count( ).sort_values(ascending=False) .head(10)
top_items.name='Count'
top_item_df = top_items.to_frame().reset_index()
top_item_df

item_id  Count
0  112921337    953
1   97655171    812
2  387911330    706
3  135104537    630
4   14087919    562
5  128186279    556
6    2217535    545
7    5685392    531
8  277922302    522
9  209323160    509

In [40]:
top_item_pie =(
    Pie(init_opts=opts.InitOpts(width='400px'))
    .add("",top_item_df.values.tolist())
    .set_global_opts(
        title_opts=opts.TitleOpts(title="Top10商品"),
        legend_opts=opts. LegendOpts(is_show=False),
    )
    .set_series_opts(label_opts=opts.LabelOpts(formatter="{d}%"))
)

top_item_pie.render_notebook()


用户行为漏斗分析

In [41]:
behavior_count = data_df.groupby('behavior_desc').size()
behavior_count

behavior_desc
加入购物车     333371
支付        110672
收藏        240919
点击       7479078
dtype: int64

In [42]:
pv = behavior_count['点击']
favor = behavior_count['收藏']
cart = behavior_count['加入购物车']
buy = behavior_count['支付']

In [43]:
#点击转化率计算
print('点击 —— 收藏转化率: {:.2%} '.format(favor / pv))
print('点击 —— 购物车转化率:{:.2%} '.format(cart / pv))
print('点击 —— 支付转化率:{:.2%}'.format(buy / pv))

点击 —— 收藏转化率: 3.22% 
点击 —— 购物车转化率:4.46% 
点击 —— 支付转化率:1.48%


In [44]:
#收藏+购物车转化率
print('收藏 —— 支付转化率:{:.2%}'.format(buy / (favor + cart)))

收藏 —— 支付转化率:19.27%


In [45]:
#可视化用户转化漏斗
behabior_category = ['点击','收藏/加入购物车','支付']
rates =[1*100,(favor + cart) / pv*100,buy / pv * 100]
format_rates = [round(r,2) for r in rates]

funnel = (
    Funnel(init_opts=opts.InitOpts(width='400px'))
    .add(
        series_name='用户行为',
        data_pair=[i for i in zip(behabior_category,format_rates)],
        tooltip_opts=opts.TooltipOpts(
            trigger='item',
            formatter='{b}: {c}%' ,
            is_show=True) ,
        label_opts=opts.LabelOpts(
            is_show=True,
            position='outside ' ),
    )
    .set_global_opts(
        title_opts=opts.TitleOpts(title='用户行为漏斗分析'),
        legend_opts=opts.LegendOpts(is_show=False)
    )
)
funnel.render_notebook()

观察结果:用户总体转化率较低，其中点击一支付转化率仅有1.48%，而点击一收藏和点击—购物车的转化率相对高一些，但也仅有3%和4%;而收藏或加入购物车的转化率为19.27%，对比点击转化率比较大的提升。
小结:用户的点击转化率总体偏低，而收藏和加购这类表现出商品兴趣的行为转化率会有较大的提升。
建议:
1.针对点击转化率的问题，一方面可以多举办限时促销活动，促使用户尽快下单，另一方面拓展站外流量来源，增加点击量;
⒉.在功能设计上，引导用户收藏和加购商品的功能，并在收藏和购物车页面配合提醒功能，以增加购买转化率。

用户价值分类

用户RFM模型分析:
R(Recency)最近一次消费时间:表示用户最近一次消费距离现在的时间。消费时间越近的客户价值越大。1年前消费过的用户肯定没有1周前消费过的用户价值大。
F (Freauency)消费频率:指用户在统计周期内购买商品的次数，经常购买的用户也就是熟客，价值肯定比偶尔来一次的客户价值大。
M(Monetary)消费金额:指用户在统计周期内消费的总金额，体现了消费者为企业创利的多少，自然是消费越多的用户价值越大。

In [46]:
#构造R值
last_time = data_df[data_df['behavior_desc']== '支付'].groupby('user_id').max()['date']
recency = (data_df['date'].max()-last_time).dt.days.copy()
median_recency = recency.median()
recency [recency <= median_recency] = 1
recency [recency > median_recency] = 0
recency.name = 'R'
recency.head(20)

user_id
4913      1
6118      1
7528      0
7591      0
12645     1
54056     0
63348     0
79824     1
88930     1
100539    1
104155    1
109103    0
113251    0
113960    1
120873    1
134658    0
151617    1
156608    0
157097    0
189833    1
Name: R, dtype: int64

In [47]:
#构建F值
frequency = data_df[data_df['behavior_desc']=='支付'].groupby('user_id')['item_id'].size()
median_freq = frequency.median()
frequency[frequency <= median_freq] = 0
frequency[frequency > median_freq] = 1
frequency.name = 'F'
frequency.head(20)

user_id
4913      0
6118      0
7528      0
7591      1
12645     0
54056     0
63348     0
79824     1
88930     1
100539    1
104155    0
109103    1
113251    0
113960    1
120873    0
134658    0
151617    1
156608    1
157097    1
189833    1
Name: F, dtype: int64

In [48]:
rfm = pd.merge(recency,frequency,on='user_id' , how='inner')
rfm = rfm[['R','F']].astype('str')
rfm['RF'] = rfm['R'] +rfm['F']
rfm.head()

R  F  RF
user_id          
4913     1  0  10
6118     1  0  10
7528     0  0  00
7591     0  1  01
12645    1  0  10

In [49]:
user_tag_dict = {'11':'重要价值客户','10':'重要发展客户','01':'重要保持客户','00':'重要挽留客户'}
rfm['user_tag'] =rfm['RF'].map(user_tag_dict)
rfm.head()

R  F  RF user_tag
user_id                   
4913     1  0  10   重要发展客户
6118     1  0  10   重要发展客户
7528     0  0  00   重要挽留客户
7591     0  1  01   重要保持客户
12645    1  0  10   重要发展客户

In [50]:
#统计分层客户数量
user_tag_count = rfm.groupby ( 'user_tag' ).size()
user_tag_count

user_tag
重要价值客户    3089
重要保持客户    1135
重要发展客户    1811
重要挽留客户    2851
dtype: int64

In [51]:
user_tag_count.index

Index(['重要价值客户', '重要保持客户', '重要发展客户', '重要挽留客户'], dtype='object', name='user_tag')

In [52]:
#可视化统计分层客户数量
user_tag_pie = (
    Pie(init_opts=opts.InitOpts(width='400px'))
    .set_global_opts(
        legend_opts=opts.LegendOpts(is_show=False)
    )
    .add(
        series_name='购买次数占比',
        data_pair=[list(z) for z in zip(user_tag_count.index.tolist(),user_tag_count.values.tolist())],
        radius=['30%', '70%'],
        label_opts=opts.LabelOpts(
            is_show=True,
            position= 'outside'),
    )
    .set_series_opts(
        tooltip_opts=opts.TooltipOpts(
            trigger='item',
            formatter='{b}: {d}%')
    )
)
user_tag_pie.render_notebook()

结论:
1.重要价值客户是最优质的用户群体，应重点关注，既要保持其粘性，又要继续引导消费，可为这类用户提供VIP服务;
2重要发展客户的特点是近期有消费但频次不高，策略是提高其消费次数，具体措施有促销活动提醒和优惠卷活动等;
3.重要保持客户的特点是消费频次高但有一段时间没有消费，策略是重新唤醒，通过APP消息推送，以及站外广告营销吸引其注意力，促进复购;
4.重要挽留客户近期没有消费且频次不高，若不加以挽留，会有流失的可能，对于这类用户一方面需要保持曝光量，持续推送活动和优惠信息，另一方面需要进—步研究其兴趣和需求，才能采取有效的运营策略。

数据大屏制作

In [53]:
page_user =(
    Page()
    .add(daily_pv_line)
    .add(hourly_pv_line)
    .add(buy_hist_bar)
)

page_user.render('./page_user.html')

page_product =(
    Page()
    .add(top_cat_pie)
    .add(top_item_pie)
)
page_product.render('./page_product.html')

page_behaviour =(
    Page()
    .add(funnel)
    .add(user_tag_pie)
)
page_behaviour .render('./page_behaviour.html')

'd:\\代码\\page_behaviour.html'